builder/build_validate.go

```diff
@@ -0,0 +1,911 @@
+}
+
+func getPkgDiffLists(fromPkgs, toPkgs map[string]bool, bundleDiff *mcaBundleDiff) diffLists {
+	addList := []string{}
+	delList := []string{}
+	modList := []string{}
+
+	for p := range toPkgs {
+		if fromPkgs[p] != false {
+			// Match
+			if isPkgMod(p, bundleDiff) {
+				modList = append(modList, p)
+			}
+		} else {
+			// Added pkg
+			addList = append(addList, p)
+		}
+	}
+
+	for p := range fromPkgs {
+		if toPkgs[p] == false {
+			// Deleted pkg
+			delList = append(delList, p)
+		}
+	}
+	return diffLists{modList: modList, addList: addList, delList: delList}
+}
+
+func getManFileDiffLists(fromFiles, toFiles map[string]*swupd.File) diffLists {
+	addList := []string{}
+	delList := []string{}
+	modList := []string{}
+
+	for f := range toFiles {
+		if fromFiles[f] != nil {
+			// Match
+			if isManFileMod(fromFiles[f], toFiles[f]) {
+				modList = append(modList, f)
+			}
+		} else {
+			// Added file
+			addList = append(addList, f)
+		}
+	}
+
+	for f := range fromFiles {
+		if toFiles[f] == nil {
+			// Deleted file
+			delList = append(delList, f)
+		}
+	}
+	return diffLists{modList: modList, addList: addList, delList: delList}
+}
+
+func isFileMod(from, to *fileInfo) bool {
+	return ((from.name != to.name) ||
+		(from.hash != to.hash) ||
+		(from.modes != to.modes) ||
+		(from.links != to.links) ||
+		(from.user != to.user) ||
+		(from.group != to.group))
+}
+
+func isPkgMod(key string, bundleDiff *mcaBundleDiff) bool {
+	// When file(s) associated with a package are modified, the package is modified
+	if (bundleDiff.pkgFileCounts[key].add + bundleDiff.pkgFileCounts[key].mod + bundleDiff.pkgFileCounts[key].del) != 0 {
+		return true
+	}
+	return false
+}
+
+func isManFileMod(from, to *swupd.File) bool {
+	return ((from.Name != to.Name) ||
+		(from.Hash != to.Hash))
+}
+
+// analyzeMcaResults compares manifest file changes against package file changes.
+// When there are inconsistencies between the manifest and package file lists,
+// a slice of error strings is returned.
+func analyzeMcaResults(results *mcaDiffResults, fromInfo, toInfo map[string]*mcaBundleInfo) ([]string, error) {
+	var errorList = []string{}
+	var tmp = []string{}
+
+	if results.bundleDiff["os-core"] == nil {
+		return nil, fmt.Errorf("os-core does not have mandatory file changes")
+	}
+
+	// The os-release, version, and versiontimestamp files are required to be
+	// modified by the os-core bundle and are not included by an rpm. Confirm
+	// that these files were modified in the manifest, but omit them from the
+	// package comparison.
+	mandatoryFiles := 0
+	for _, f := range results.bundleDiff["os-core"].manFileDiffs.modList {
+		switch f {
+		case "/usr/lib/os-release":
+			mandatoryFiles++
+		case "/usr/share/clear/version":
+			mandatoryFiles++
+		case "/usr/share/clear/versionstamp":
+			mandatoryFiles++
+		default:
+			tmp = append(tmp, f)
+		}
+	}
+	results.bundleDiff["os-core"].manFileDiffs.modList = tmp
+
+	if mandatoryFiles != 3 {
```

Comment:
These file exceptions will be refactored to handle +10, +20, and different formats.